README.MD
# Справка
## HTML
Работа с HTML.

## Использование
### Получение шаблона регулярного выражения
#### Область CDATA
```java
Regexp.CDATA;
```

#### Самозакрывающийся HTML-тег
```java
Regexp.SELF_CLOSING_TAG;
```

#### Открывающий HTML-тег
```java
Regexp.OPEN_TAG;
```

#### Закрывающий HTML-тег
```java
Regexp.CLOSE_TAG;
```

#### HTML-комментарий
```java
Regexp.COMMENT;
```

#### Тип HTML-документа
```java
Regexp.DOCTYPE;
```

#### HTML-сущность в виде имени
```java
Regexp.ENTITY_NAME;
```

#### HTML-сущность в виде unicode
```java
Regexp.ENTITY_UNICODE;
```

#### HTML-сущность в виде числа
```java
Regexp.ENTITY_NUMERIC;
```

### Проверка
#### Проверка HTML-комментария
```java
String text = "<!-- Комментарий -->";

Html html = new Html(text);
html.isComment();
```

Результат выполнения:
```
true
```

#### Проверка HTML-сущности
```java
String text = "&asymp;";

Html html = new Html(text);
html.isEntity();
```

Результат выполнения:
```
true
```

#### Проверка HTML-сущности в виде имени
```java
String text = "&asymp;";

Html html = new Html(text);
html.isEntityName();
```

Результат выполнения:
```
true
```

#### Проверка HTML-сущности в виде числа
```java
String text = "&38;";

Html html = new Html(text);
html.isEntityNumeric();
```

Результат выполнения:
```
true
```

#### Проверка HTML-сущности в виде unicode
```java
String text = "&#X22C8;";

Html html = new Html(text);
html.isEntityUnicode();
```

Результат выполнения:
```
true
```

#### Проверка типа HTML-документа
```java
String text = "<!DOCTYPE html>";

Html html = new Html(text);
html.isDoctype();
```

Результат выполнения:
```
true
```

#### Проверка HTML-тега
```java
String text = "<b class='xa-xa'>";

Html html = new Html(text);
html.isTag();
```

Результат выполнения:
```
true
```

#### Проверка открывающего HTML-тега
```java
String text = "<input type='checkbox'>";

Html html = new Html(text);
html.isOpenTag();
```

Результат выполнения:
```
true
```

#### Проверка самозакрывающегося HTML-тега
```java
String text = "<input type='checkbox' />";

Html html = new Html(text);
html.isSelfClosingTag();
```

Результат выполнения:
```
true
```

#### Проверка закрывающего HTML-тега
```java
String text = "</form>";

Html html = new Html(text);
html.isCloseTag();
```

Результат выполнения:
```
true
```

#### Проверка области CData
```java
String text = "<![CDATA[CDATA область]]>";

Html html = new Html(text);
html.isCData();
```

Результат выполнения:
```
true
```

### Поиск
#### Поиск HTML-комментариев
```java
String text = """
	Песня без слов, ночь без сна
	<!--Всё в свое время, зима и весна-->
	Каждой звезде свой неба кусок
	Каждому морю дождя глоток
	Каждому <!-- яблоку место упасть
	Каждому вору возможность украсть
	Каждой собаке палку и кость-->
	И каждому волку зубы и злость
	""";

Html html = new Html(text);
html.findComments();
```

Результат выполнения:
```
[
	"<!--Всё в свое время, зима и весна-->",
	"<!-- яблоку место упасть\nКаждому вору возможность украсть\nКаждой собаке палку и кость-->"
]
```

#### Поиск областей CDATA
```java
String text = """
	<![CDATA[Здесь не понятно, где лицо, а где рыло,]]>
	И не понятно, где пряник, где плеть.
	<![CDATA[Здесь в сено не втыкаются вилы,
	А рыба проходит сквозь сеть.
	И не ясно, где море, где суша,
	Где золото, а где медь.]]>
	Что построить, и что разрушить,
	И кому, и зачем здесь петь?
	""";

Html html = new Html(text);
html.findCData();
```

Результат выполнения:
```
[
	"<![CDATA[Здесь не понятно, где лицо, а где рыло,]]>",
	"<![CDATA[Здесь в сено не втыкаются вилы,\nА рыба проходит сквозь сеть.\nИ не ясно, где море, где суша,\nГде золото, а где медь.]]>"
]
```

#### Поиск HTML-сущностей
```java
String text = """
	Как много веселых ребят,
	И все делают &frac14; велосипед,
	А один из них как-нибудь утром придумает порох.
	Ну а я здесь сижу &#44;без тебя,&amp;
	Мне до этих ребят дела нет,
	Лишь окурки лежат на полу, да мусора ворох.&#169;
	""";

Html html = new Html(text);
html.findEntities();
```

Результат выполнения:
```
[
	"&frac14;",
	"&amp;",
	"&#44;",
	"&#169;"
]
```

#### Поиск HTML-сущностей в виде имени
```java
String text = """
	Как много веселых ребят,
	И все делают &frac14; велосипед,
	А один из них как-нибудь утром придумает порох.
	Ну а я здесь сижу без тебя,&amp;
	Мне до этих ребят дела нет,
	Лишь окурки лежат на полу, да мусора ворох.
	""";

Html html = new Html(text);
html.findEntitiesName();
```

Результат выполнения:
```
[
	"&frac14;",
	"&amp;"
]
```

#### Поиск HTML-сущностей в виде числа
```java
String text = """
	В моем доме не видно стены,
	В моем небе &#44;не видно луны.
	Я слеп, но я вижу тебя,
	Я глух, но я слышу тебя.
	Я не сплю, &#169; но я вижу сны,
	Здесь нет моей вины,
	Я нем, но ты слышишь меня,
	И этим мы сильны.
	""";

Html html = new Html(text);
html.findEntitiesNumeric();
```

Результат выполнения:
```
[
	"&#44;",
	"&#169;"
]
```

#### Поиск HTML-сущностей в виде unicode
```java
String text = """
	Ночь, день -
	Спать &#X22C8;лень.
	Есть дым -&#x224E;
	Чёрт с ним.
	Сна нет -
	Есть сон лет.
	Кино -
	Кончилось давно.
	""";

Html html = new Html(text);
html.findEntitiesUnicode();
```

Результат выполнения:
```
[
	"&#X22C8;",
	"&#x224E;"
]
```

#### Поиск типа HTML-документа
```java
String text = """
	<!DOCTYPE html>Они говорят: им нельзя рисковать,
	Потому что у них есть дом,
	В доме горит свет.
	И я не знаю точно, <!DOCTYPE html PUBLIC '-//W3C//DTD XHTML 1.0 Strict//EN'
	'http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd'> кто из нас прав,
	Меня ждет на улице дождь,
	Их ждет дома обед.
	""";

Html html = new Html(text);
html.findDoctype();
```

Результат выполнения:
```
[
	"<!DOCTYPE html>",
	"<!DOCTYPE html PUBLIC '-//W3C//DTD XHTML 1.0 Strict//EN'\n'http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd'>"
]
```

#### Поиск HTML-тегов
```java
String text = """
	Солдат шел по улице домой
	И <b>увидел этих ребят.</b >
	"Кто ваша мама, ребята?" -
	<meta charset='UTF-8'>Спросил у ребят солдат.
	""";

Html html = new Html(text);
html.findTags();
```

Результат выполнения:
```
[
	"<b>",
	"</b >",
	"<meta charset='UTF-8'>"
]
```

#### Поиск открывающих HTML-тегов
```java
String text = """
	Солдат шел по улице домой
	И <b>увидел этих ребят.
	"Кто ваша мама, ребята?" -
	<meta charset='UTF-8'>Спросил у ребят солдат.
	""";

Html html = new Html(text);
html.findOpenTags();
```

Результат выполнения:
```
[
	"<b>",
	"<meta charset='UTF-8'>"
]
```

#### Поиск закрывающих HTML-тегов
```java
String text = """
	О-o, это странное место Камчатка,
	О-o, это сладкое</form> слово "Камчатка".
	Но на этой земле я не вижу тебя,
	Я не вижу твоих кораблей,
	Я не вижу реки, я не вижу моста,
	Ну и пусть...</b >
	""";

Html html = new Html(text);
html.findCloseTags();
```

Результат выполнения:
```
[
	"</form>",
	"</b >"
]
*/

```
#### Поиск самозакрывающихся HTML-тегов
```java
String text = """
	У меня есть дом, только нет ключей,
	У меня есть солнце, но оно среди туч,<br/>
	Есть голова<meta charset='UTF-8'/>, только нет плечей,
	Но я вижу, как тучи режут солнечный луч.
	У меня есть слово, но в нем нет букв,
	У меня есть лес, но нет топоров,
	У меня есть время, но нет сил ждать,
	И есть еще ночь, но в ней нет снов.
	""";

Html html = new Html(text);
html.findSelfClosingTags();
```

Результат выполнения:
```
[
	"<br/>",
	"<meta charset='UTF-8'/>"
]
```

### Преобразование
#### Удаление HTML-комментариев
```java
String text = """
	<html><!-- Комментарий --></html>
	<!--
		Комментарий
		с переносом
	-->
	""";

Html html = new Html(text);
html.deleteComments();

```

Результат выполнения:
```
<html></html>
```

#### Удаление разделов CDATA
```java
String text = """
	<html>
	<![CDATA[x<y]]>Текст<![CDATA[
	cdata ->
	text
	]]>
	</html>
	""";

Html html = new Html(text);
html.deleteCData();
```

Результат выполнения:
```
<html>
Текст
</html>
```

#### Удаление типа документа
```java
String text = """
	<!DOCTYPE html>
	<b><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
		"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"></b>
	""";

Html html = new Html(text);
html.deleteDoctype();
```

Результат выполнения:
```
<b></b>
```

#### Удаление HTML-тега
```java
String text = """
	text <a href="/link" target='_blank'>ссылка</a> Текст
	""";

Html html = new Html(text);
html.deleteTag(Tag.A);
```

Результат выполнения:
```
text ссылка Текст
```

#### Удаление HTML-тега указанного типа
```java
String text = """
	text <a href="/link" target='_blank'>ссылка</a> Текст
	""";

Html html = new Html(text);
html.deleteTag(Tag.A, TagType.OPEN);
```

Результат выполнения:
```
text ссылка</a> Текст
```

#### Удаление всех HTML-тегов
```java
String text = """
	<b>text</b> <a href="/link" target='_blank'>ссылка</a> Текст <br/>
	""";

Html html = new Html(text);
html.deleteTags();
```

Результат выполнения:
```
text ссылка Текст
```

#### Удаление всех HTML-тегов указанного типа
```java
String text = """
	<b>text</b> <a href="/link" target='_blank'>ссылка</a> Текст <br/>
	""";

Html html = new Html(text);
html.deleteTags(TagType.OPEN);
```

Результат выполнения:
```
text</b> ссылка</a> Текст <br/>
```

#### Удаление всех HTML-тегов с указанием исключений
```java
String text = """
	<b>text</b> <a href="/link" target='_blank'>ссылка</a> Текст <br/>
	""";

Html html = new Html(text);
html.deleteTags(Set.of(Tag.BR));
```

Результат выполнения:
```
text ссылка Текст <br/>
```

#### Удаление всех HTML-тегов с указанного типа и исключений
```java
String text = """
	<b>text</b> <a href="/link" target='_blank'>ссылка</a> Текст <br/>
	""";

Html html = new Html(text);
html.deleteTags(TagType.OPEN, Set.of(Tag.B));
```

Результат выполнения:
```
<b>text</b> ссылка</a> Текст <br/>
```

#### Удаление HTML-сущности
```java
String text = """
	AC&sol;DC
	&lt;&#60;&#x3C;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
	""";

Html html = new Html(text);
html.deleteEntity(Entity.LT);
```

Результат выполнения:
```
AC&sol;DC
переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
```

#### Удаление HTML-сущности указанного типа
```java
String text = """
	AC&sol;DC
	&lt;&#60;&#x3C;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
	""";

Html html = new Html(text);
html.deleteEntity(Entity.LT, EntityType.NUMERIC);
```

Результат выполнения:
```
AC&sol;DC
&lt;&#x3C;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
```

#### Удаление всех HTML-сущностей
```java
String text = """
	AC&sol;DC
	&lt;&#60;&#x3C;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
	""";

Html html = new Html(text);
html.deleteEntities();
```

Результат выполнения:
```
ACDC
переменный токпостоянный ток
```

#### Удаление всех HTML-сущностей с указанием исключений
```java
	String text = """
	AC&sol;DC
	&lt;&#60;&#x3C;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
	""";

Html html = new Html(text);
html.deleteEntities(Set.of(Entity.LT, Entity.GT));
```

Результат выполнения:
```
ACDC
&lt;&#60;&#x3C;переменный токпостоянный ток&#62;&#X0003E;&gt;
```

#### Преобразование специальных символов в соответствующие HTML-сущности с указанием типа
```java
String text = """
	AC/DC
	<<<переменный ток/постоянный ток>>>;
	""";

Html html = new Html(text);
html.encodeEntity(Entity.LT, EntityType.NAME);
```

Результат выполнения:
```
AC/DC
&lt;&lt;&lt;переменный ток/постоянный ток>>>
```

#### Преобразование всех специальных символов в соответствующие HTML-сущности с указанием типа
```java
String text = """
	AC/DC
	<<<переменный ток/постоянный ток>>>;
	""";

Html html = new Html(text);
html.encodeEntities(EntityType.NAME);
```

Результат выполнения:
```
AC&sol;DC
&lt;&lt;&lt;переменный ток&sol;постоянный ток&gt;&gt;&gt;
```

#### Преобразование всех специальных символов в соответствующие HTML-сущности с указанием типа и исключений
```java
String text = """
	AC/DC
	<<<переменный ток/постоянный ток>>>;
	""";

Html html = new Html(text);
html.encodeEntities(EntityType.NAME, Set.of(Entity.GT));
```

Результат выполнения:
```
AC&sol;DC
&lt;&lt;&lt;переменный ток&sol;постоянный ток>>>
```

#### Преобразование HTML-сущности в соответствующий специальный символ
```java
String text = """
	AC&sol;DC
	&lt;&lt;&lt;переменный ток&sol;постоянный ток&gt;&gt;&gt;
	""";

Html html = new Html(text);
html.decodeEntity(Entity.GT);
```

Результат выполнения:
```
AC&sol;DC
&lt;&lt;&lt;переменный ток&sol;постоянный ток>>>
```

#### Преобразование HTML-сущности в соответствующий специальный символ с указанием типа
```java
String text = """
	AC&sol;DC
	&lt;&lt;&lt;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
	""";

Html html = new Html(text);
html.decodeEntity(Entity.GT, EntityType.NAME);
```

Результат выполнения:
```
AC&sol;DC
&lt;&lt;&lt;переменный ток&sol;постоянный ток&#62;&#X0003E;>
```

#### Преобразование всех HTML-сущностей в соответствующие специальные символы
```java
String text = """
	AC&sol;DC
	&lt;&lt;&lt;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
	""";

Html html = new Html(text);
html.decodeEntities();
```

Результат выполнения:
```
AC/DC
<<<переменный ток/постоянный ток>>>
```

#### Преобразование всех HTML-сущностей в соответствующие специальные символы с указанием типа
```java
String text = """
	AC&sol;DC
	&lt;&lt;&lt;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
	""";

Html html = new Html(text);
html.decodeEntities(EntityType.NAME);
```

Результат выполнения:
```
AC/DC
<<<переменный ток/постоянный ток&#62;&#X0003E;>
```

#### Преобразование всех HTML-сущностей в соответствующие специальные символы с указанием исключений
```java
String text = """
	AC&sol;DC
	&lt;&lt;&lt;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
	""";

Html html = new Html(text);
html.decodeEntities(Set.of(Entity.LT, Entity.GT));
```

Результат выполнения:
```
AC/DC
&lt;&lt;&lt;переменный ток/постоянный ток&#62;&#X0003E;&gt;
```

#### Преобразование всех HTML-сущностей в соответствующие специальные символы с указанием типа и исключений
```java
String text = """
	AC&sol;DC
	&lt;&lt;&lt;переменный ток&sol;постоянный ток&#62;&#X0003E;&gt;
	""";

Html html = new Html(text);
html.decodeEntities(EntityType.NAME, Set.of(Entity.LT));
```

Результат выполнения:
```
AC/DC
&lt;&lt;&lt;переменный ток/постоянный ток&#62;&#X0003E;>
```


CHANGELOG.MD
# Журнал изменений
## vX.X.X - XX.XX.XXXX
### Изменено
* Переименован пакет resources -> types.

## v1.4.0 - 08.07.2023
### Изменено
* Изменена версия Java на 17.

## v1.3.0 - 14.05.2023
### Добавлено
* Добавлен метод Hypertext.isEntity().
* Добавлен метод Hypertext.isTag().
* Добавлен метод Hypertext.findEntities().
* Добавлен метод Hypertext.findTags().

### Изменено
* Переименован интерфейс Hipertext -> Hypertext
* Переименован метод Hypertext.findOpenTag() -> Hypertext.findOpenTags().
* Переименован метод Hypertext.findSelfClosingTag() -> Hypertext.findSelfClosingTags().
* Переименован метод Hypertext.findCloseTag() -> Hypertext.findCloseTags().
* Переименован метод Hypertext.findComment() -> Hypertext.findComments().
* Переименован метод Hypertext.findEntity*() -> Hypertext.findEntities*().

### Исправлено
* Исправлен баг в регулярном выражении Regexp.OPEN_TAG.

## v1.2.0 - 20.01.2023
### Добавлено
* Добавлен AUTHOR.MD.
* Добавлен INSTALL.MD.
* Добавлен CHANGELOG.MD.
* Добавлено перечисление Regexp.
* Добавлен метод Hypertext.isCData().
* Добавлен метод Hypertext.isComment().
* Добавлен метод Hypertext.isDoctype().
* Добавлен метод Hypertext.isEndTag().
* Добавлен метод Hypertext.isEntityName().
* Добавлен метод Hypertext.isEntityNumeric().
* Добавлен метод Hypertext.isEntityUnicode().
* Добавлен метод Hypertext.isSelfClosingTag().
* Добавлен метод Hypertext.isStartTag().
* Добавлен метод Hypertext.findCData().
* Добавлен метод Hypertext.findComment().
* Добавлен метод Hypertext.findDoctype().
* Добавлен метод Hypertext.findEndTag().
* Добавлен метод Hypertext.findEntityName().
* Добавлен метод Hypertext.findEntityNumeric().
* Добавлен метод Hypertext.findEntityUnicode().
* Добавлен метод Hypertext.findSelfClosingTags().
* Добавлен метод Hypertext.findStartTag().

### Изменено
* Обновлён README.MD.
* Переименовано Regexp.START_TAG -> Regexp.OPEN_TAG.
* Переименовано Regexp.END_TAG -> Regexp.CLOSE_TAG.
* Переименовано TagType.START -> TagType.OPEN.
* Переименовано TagType.END -> TagType.CLOSE.
* Переименован метод Hypertext.isEndTag() -> Hypertext.isCloseTag().
* Переименован метод Hypertext.findEndTag() -> Hypertext.findCloseTags().
* Переименован метод Hypertext.isStartTag() -> Hypertext.isOpenTag().
* Переименован метод Hypertext.findStartTag() -> Hypertext.findOpenTag().

## v1.1.0 - 16.10.2022
### Добавлено
* Добавлен плагин maven-assembly в pom.xml.
* Добавлен плагин maven-compiler в pom.xml.
* Добавлен плагин maven-dependency в pom.xml.
* Добавлен плагин maven-javadoc в pom.xml.
* Добавлен плагин maven-source в pom.xml.

### Изменено
* Изменён комментарий для Entity.Entity(Character, String, String, String).
* Изменён комментарий для Entity.getEntity(EntityType).


src/main/java/dev/kalenchukov/html/types/Tag.java
/*
 * Copyright © 2022-2023 Алексей Каленчуков
 * GitHub: https://github.com/kalenchukov
 * E-mail: mailto:aleksey.kalenchukov@yandex.ru
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dev.kalenchukov.html.types;

import org.jetbrains.annotations.NotNull;

/**
 * Перечисление HTML-тегов.
 *
 * @author Алексей Каленчуков
 */
public enum Tag
{
	/**
	 * Тег &#60;HTML&#62;.
	 */
	HTML("html"),

	/**
	 * Тег &#60;HEAD&#62;.
	 */
	HEAD("head"),

	/**
	 * Тег &#60;BODY&#62;.
	 */
	BODY("body"),

	/**
	 * Тег &#60;DIV&#62;.
	 */
	DIV("div"),

	/**
	 * Тег &#60;ASIDE&#62;.
	 */
	ASIDE("aside"),

	/**
	 * Тег &#60;FOOTER&#62;.
	 */
	FOOTER("footer"),

	/**
	 * Тег &#60;FORM&#62;.
	 */
	FORM("form"),

	/**
	 * Тег &#60;FRAME&#62;.
	 */
	FRAME("frame"),

	/**
	 * Тег &#60;HEADER&#62;.
	 */
	HEADER("header"),

	/**
	 * Тег &#60;IFRAME&#62;.
	 */
	IFRAME("iframe"),

	/**
	 * Тег &#60;MENU&#62;.
	 */
	MENU("menu"),

	/**
	 * Тег &#60;META&#62;.
	 */
	META("meta"),

	/**
	 * Тег &#60;SCRIPT&#62;.
	 */
	SCRIPT("script"),

	/**
	 * Тег &#60;SECTION&#62;.
	 */
	SECTION("section"),

	/**
	 * Тег &#60;TABLE&#62;.
	 */
	TABLE("table"),

	/**
	 * Тег &#60;TBODY&#62;.
	 */
	TBODY("tbody"),

	/**
	 * Тег &#60;A&#62;.
	 */
	A("a"),

	/**
	 * Тег &#60;ABBR&#62;.
	 */
	ABBR("abbr"),

	/**
	 * Тег &#60;ACRONYM&#62;.
	 */
	ACRONYM("acronym"),

	/**
	 * Тег &#60;ADDRESS&#62;.
	 */
	ADDRESS("address"),

	/**
	 * Тег &#60;APPLET&#62;.
	 */
	APPLET("applet"),

	/**
	 * Тег &#60;AREA&#62;.
	 */
	AREA("area"),

	/**
	 * Тег &#60;ARTICLE&#62;.
	 */
	ARTICLE("article"),

	/**
	 * Тег &#60;AUDIO&#62;.
	 */
	AUDIO("audio"),

	/**
	 * Тег &#60;B&#62;.
	 */
	B("b"),

	/**
	 * Тег &#60;BASE&#62;.
	 */
	BASE("base"),

	/**
	 * Тег &#60;BDI&#62;.
	 */
	BDI("bdi"),

	/**
	 * Тег &#60;BASEFONT&#62;.
	 */
	BASEFONT("basefont"),

	/**
	 * Тег &#60;BDO&#62;.
	 */
	BDO("bdo"),

	/**
	 * Тег &#60;BGSOUND&#62;.
	 */
	BGSOUND("bgsound"),

	/**
	 * Тег &#60;BIG&#62;.
	 */
	BIG("big"),

	/**
	 * Тег &#60;BLINK&#62;.
	 */
	BLINK("blink"),

	/**
	 * Тег &#60;BLOCKQUOTE&#62;.
	 */
	BLOCKQUOTE("blockquote"),

	/**
	 * Тег &#60;BR&#62;.
	 */
	BR("br"),

	/**
	 * Тег &#60;BUTTON&#62;.
	 */
	BUTTON("button"),

	/**
	 * Тег &#60;CANVAS&#62;.
	 */
	CANVAS("canvas"),

	/**
	 * Тег &#60;CAPTION&#62;.
	 */
	CAPTION("caption"),

	/**
	 * Тег &#60;CENTER&#62;.
	 */
	CENTER("center"),

	/**
	 * Тег &#60;CITE&#62;.
	 */
	CITE("cite"),

	/**
	 * Тег &#60;CODE&#62;.
	 */
	CODE("code"),

	/**
	 * Тег &#60;COMMENT&#62;.
	 */
	COMMENT("comment"),

	/**
	 * Тег &#60;COL&#62;.
	 */
	COL("col"),

	/**
	 * Тег &#60;COLGROUP&#62;.
	 */
	COLGROUP("colgroup"),

	/**
	 * Тег &#60;COMMAND&#62;.
	 */
	COMMAND("command"),

	/**
	 * Тег &#60;DATALIST&#62;.
	 */
	DATALIST("datalist"),

	/**
	 * Тег &#60;DD&#62;.
	 */
	DD("dd"),

	/**
	 * Тег &#60;DEL&#62;.
	 */
	DEL("del"),

	/**
	 * Тег &#60;DIR&#62;.
	 */
	DIR("dir"),

	/**
	 * Тег &#60;DETAILS&#62;.
	 */
	DETAILS("details"),

	/**
	 * Тег &#60;DFN&#62;.
	 */
	DFN("dfn"),

	/**
	 * Тег &#60;DL&#62;.
	 */
	DL("dl"),

	/**
	 * Тег &#60;DT&#62;.
	 */
	DT("dt"),

	/**
	 * Тег &#60;EM&#62;.
	 */
	EM("em"),

	/**
	 * Тег &#60;EMBED&#62;.
	 */
	EMBED("embed"),

	/**
	 * Тег &#60;FIELDSET&#62;.
	 */
	FIELDSET("fieldset"),

	/**
	 * Тег &#60;FIGCAPTION&#62;.
	 */
	FIGCAPTION("figcaption"),

	/**
	 * Тег &#60;FIGURE&#62;.
	 */
	FIGURE("figure"),

	/**
	 * Тег &#60;FONT&#62;.
	 */
	FONT("font"),

	/**
	 * Тег &#60;FRAMESET&#62;.
	 */
	FRAMESET("frameset"),

	/**
	 * Тег &#60;H1&#62;.
	 */
	H1("h1"),

	/**
	 * Тег &#60;H2&#62;.
	 */
	H2("h2"),

	/**
	 * Тег &#60;H3&#62;.
	 */
	H3("h3"),

	/**
	 * Тег &#60;H4&#62;.
	 */
	H4("h4"),

	/**
	 * Тег &#60;H5&#62;.
	 */
	H5("h5"),

	/**
	 * Тег &#60;H6&#62;.
	 */
	H6("h6"),

	/**
	 * Тег &#60;HGROUP&#62;.
	 */
	HGROUP("hgroup"),

	/**
	 * Тег &#60;HR&#62;.
	 */
	HR("hr"),

	/**
	 * Тег &#60;I&#62;.
	 */
	I("i"),

	/**
	 * Тег &#60;ISINDEX&#62;.
	 */
	ISINDEX("isindex"),

	/**
	 * Тег &#60;IMG&#62;.
	 */
	IMG("img"),

	/**
	 * Тег &#60;INPUT&#62;.
	 */
	INPUT("input"),

	/**
	 * Тег &#60;INS&#62;.
	 */
	INS("ins"),

	/**
	 * Тег &#60;KBD&#62;.
	 */
	KBD("kbd"),

	/**
	 * Тег &#60;KEYGEN&#62;.
	 */
	KEYGEN("keygen"),

	/**
	 * Тег &#60;LABEL&#62;.
	 */
	LABEL("label"),

	/**
	 * Тег &#60;LEGEND&#62;.
	 */
	LEGEND("legend"),

	/**
	 * Тег &#60;LI&#62;.
	 */
	LI("li"),

	/**
	 * Тег &#60;LINK&#62;.
	 */
	LINK("link"),

	/**
	 * Тег &#60;LISTING&#62;.
	 */
	LISTING("listing"),

	/**
	 * Тег &#60;MAIN&#62;.
	 */
	MAIN("main"),

	/**
	 * Тег &#60;MAP&#62;.
	 */
	MAP("map"),

	/**
	 * Тег &#60;MARK&#62;.
	 */
	MARK("mark"),

	/**
	 * Тег &#60;MARQUEE&#62;.
	 */
	MARQUEE("marquee"),

	/**
	 * Тег &#60;MATH&#62;.
	 */
	MATH("math"),

	/**
	 * Тег &#60;METER&#62;.
	 */
	METER("meter"),

	/**
	 * Тег &#60;NAV&#62;.
	 */
	NAV("nav"),

	/**
	 * Тег &#60;NEXTID&#62;.
	 */
	NEXTID("nextid"),

	/**
	 * Тег &#60;NOBR&#62;.
	 */
	NOBR("nobr"),

	/**
	 * Тег &#60;NOEMBED&#62;.
	 */
	NOEMBED("noembed"),

	/**
	 * Тег &#60;NOFRAMES&#62;.
	 */
	NOFRAMES("noframes"),

	/**
	 * Тег &#60;NOSCRIPT&#62;.
	 */
	NOSCRIPT("noscript"),

	/**
	 * Тег &#60;OBJECT&#62;.
	 */
	OBJECT("object"),

	/**
	 * Тег &#60;OL&#62;.
	 */
	OL("ol"),

	/**
	 * Тег &#60;OPTGROUP&#62;.
	 */
	OPTGROUP("optgroup"),

	/**
	 * Тег &#60;OPTION&#62;.
	 */
	OPTION("option"),

	/**
	 * Тег &#60;OUTPUT&#62;.
	 */
	OUTPUT("output"),

	/**
	 * Тег &#60;P&#62;.
	 */
	P("p"),

	/**
	 * Тег &#60;PARAM&#62;.
	 */
	PARAM("param"),

	/**
	 * Тег &#60;PLAINTEXT&#62;.
	 */
	PLAINTEXT("plaintext"),

	/**
	 * Тег &#60;PRE&#62;.
	 */
	PRE("pre"),

	/**
	 * Тег &#60;PROGRESS&#62;.
	 */
	PROGRESS("progress"),

	/**
	 * Тег &#60;Q&#62;.
	 */
	Q("q"),

	/**
	 * Тег &#60;RP&#62;.
	 */
	RP("rp"),

	/**
	 * Тег &#60;RT&#62;.
	 */
	RT("rt"),

	/**
	 * Тег &#60;RUBY&#62;.
	 */
	RUBY("ruby"),

	/**
	 * Тег &#60;S&#62;.
	 */
	S("s"),

	/**
	 * Тег &#60;SMALL&#62;.
	 */
	SMALL("small"),

	/**
	 * Тег &#60;SAMP&#62;.
	 */
	SAMP("samp"),

	/**
	 * Тег &#60;SELECT&#62;.
	 */
	SELECT("select"),

	/**
	 * Тег &#60;SOURCE&#62;.
	 */
	SOURCE("source"),

	/**
	 * Тег &#60;SPACER&#62;.
	 */
	SPACER("spacer"),

	/**
	 * Тег &#60;SPAN&#62;.
	 */
	SPAN("span"),

	/**
	 * Тег &#60;STRIKE&#62;.
	 */
	STRIKE("strike"),

	/**
	 * Тег &#60;STRONG&#62;.
	 */
	STRONG("strong"),

	/**
	 * Тег &#60;STYLE&#62;.
	 */
	STYLE("style"),

	/**
	 * Тег &#60;SUB&#62;.
	 */
	SUB("sub"),

	/**
	 * Тег &#60;SUMMARY&#62;.
	 */
	SUMMARY("summary"),

	/**
	 * Тег &#60;SUP&#62;.
	 */
	SUP("sup"),

	/**
	 * Тег &#60;SVG&#62;.
	 */
	SVG("svg"),

	/**
	 * Тег &#60;TD&#62;.
	 */
	TD("td"),

	/**
	 * Тег &#60;TEXTAREA&#62;.
	 */
	TEXTAREA("textarea"),

	/**
	 * Тег &#60;TFOOT&#62;.
	 */
	TFOOT("tfoot"),

	/**
	 * Тег &#60;TH&#62;.
	 */
	TH("th"),

	/**
	 * Тег &#60;THEAD&#62;.
	 */
	THEAD("thead"),

	/**
	 * Тег &#60;TIME&#62;.
	 */
	TIME("time"),

	/**
	 * Тег &#60;TITLE&#62;.
	 */
	TITLE("title"),

	/**
	 * Тег &#60;TR&#62;.
	 */
	TR("tr"),

	/**
	 * Тег &#60;TRACK&#62;.
	 */
	TRACK("track"),

	/**
	 * Тег &#60;TT&#62;.
	 */
	TT("tt"),

	/**
	 * Тег &#60;U&#62;.
	 */
	U("u"),

	/**
	 * Тег &#60;UL&#62;.
	 */
	UL("ul"),

	/**
	 * Тег &#60;VAR&#62;.
	 */
	VAR("var"),

	/**
	 * Тег &#60;VIDEO&#62;.
	 */
	VIDEO("video"),

	/**
	 * Тег &#60;WBR&#62;.
	 */
	WBR("wbr"),

	/**
	 * Тег &#60;XMP&#62;.
	 */
	XMP("xmp");

	/**
	 * Название HTML-тег.
	 */
	@NotNull
	private final String name;

	/**
	 * Конструктор для {@code Tag}.
	 *
	 * @param name название HTML-тега.
	 */
	Tag(@NotNull final String name)
	{
		this.name = name;
	}

	/**
	 * Возвращает название HTML-тега.
	 *
	 * @return название HTML-тега.
	 */
	@NotNull
	public String getName()
	{
		return this.name;
	}
}
